### PR TITLE
Rake task to archive users that are abandoned

### PR DESCRIPTION
**Pull request by slawosz**
### Context

- Ticket: 
https://dfedigital.atlassian.net/browse/CPDNPQ-1258

### Changes proposed in this pull request
Rake task and filter to find users that needs to be archived

### CODE REVIEW

File: db/schema.rb
Comment (tonyheadford) [inline db/schema.rb:1]: Can you take this out please

File: app/models/user.rb [hunk 1]
```diff
@@ -224,6 +224,26 @@ def archived?
...
+    ids_to_exclude << DeliveryPartnerProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << FinanceProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << InductionCoordinatorProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << LeadProviderProfile.where(user_id: ids).pluck(:user_id)
+
+    ids = ids - ids_to_exclude.flatten.uniq
+    User.where(id: ids)
+  end
+
+
```
Comment (tonyheadford) [inline app/models/user.rb:246]: Could we move this to a service class instead?  Would feel more natural to me to invoke this via a service but what do you think?

File: app/services/deduplication_service.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class DeduplicationService
```
Comment (tonyheadford) [inline app/services/deduplication_service.rb:3]:
```suggestion
class DeduplicationService < ::BaseService
```

File: app/services/deduplication_service.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    ids_to_exclude << AppropriateBodyProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << DeliveryPartnerProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << FinanceProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << InductionCoordinatorProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << LeadProviderProfile.where(user_id: ids).pluck(:user_id)
+
+    ids -= ids_to_exclude.flatten.uniq
+    User.where(id: ids)
+  end
+end
```
Comment (tonyheadford) [inline app/services/deduplication_service.rb:22]: Just to follow our usual service pattern

```suggestion
class DeduplicationService < ::BaseService
  def call
    select_users_to_archive.each(&:archive!)
  end
private
  def select_users_to_archive
    ids = User.left_joins(:participant_identities).where(participant_identities: { user_id: nil }).pluck(:id)

    ids_to_exclude = []
    ids_to_exclude << AdminProfile.where(user_id: ids).pluck(:user_id)
    ids_to_exclude << AppropriateBodyProfile.where(user_id: ids).pluck(:user_id)
    ids_to_exclude << DeliveryPartnerProfile.where(user_id: ids).pluck(:user_id)
    ids_to_exclude << FinanceProfile.where(user_id: ids).pluck(:user_id)
    ids_to_exclude << InductionCoordinatorProfile.where(user_id: ids).pluck(:user_id)
    ids_to_exclude << LeadProviderProfile.where(user_id: ids).pluck(:user_id)

    ids -= ids_to_exclude.flatten.uniq
    User.where(id: ids)
  end
end
```

File: lib/tasks/deduping.rake [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+namespace :deduping do
+  desc "Archive email of all users that are eligible for deduping (no participant identity associated)"
+  task dedup_users: :environment do
+    DeduplicationService.dedup_users!
```
Comment (tonyheadford) [inline lib/tasks/deduping.rake:6]:
```suggestion
    DeduplicationService.call
```

File: spec/services/deduplication_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    it "includes deletable" do
+      expect(users_to_delete).to include(deletable_user)
+    end
+  end
+
+  describe ".dedup_users" do
+    it "performs deduplication (archiving) users" do
+      expect { DeduplicationService.dedup_users! }.to change { deletable_user.reload.email }.from("user@example.org").to("user#{deletable_user.id}@example.org")
```
Comment (slawosz) [inline spec/services/deduplication_service_spec.rb:60]:
```suggestion
      expect { DeduplicationService.call }.to change { deletable_user.reload.email }.from("user@example.org").to("user#{deletable_user.id}@example.org")
```

File: app/services/deduplication_service.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  def select_users_to_archive
+    ids = User.left_joins(:participant_identities).where(participant_identities: { user_id: nil }).pluck(:id)
+
+    ids_to_exclude = []
+    ids_to_exclude << AdminProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << AppropriateBodyProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << DeliveryPartnerProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << FinanceProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << InductionCoordinatorProfile.where(user_id: ids).pluck(:user_id)
+    ids_to_exclude << LeadProviderProfile.where(user_id: ids).pluck(:user_id)
```
Comment (cwrw) [inline app/services/deduplication_service.rb:17]: could we also check that users aren't linked to declarations, also teacher profiles? As those might be edge cases that haven't been linked up properly?

Comment (slawosz) [reply app/services/deduplication_service.rb:19]: @tonyheadford I think we are allowing to dedup users who has teacher profiles?

Comment (tonyheadford) [reply app/services/deduplication_service.rb:19]: Yes some deduped orphan users will definitely have teacher profiles too @slawosz

**Review by tonyheadford — APPROVED** [review]
LGTM!

**Pull request merged**